src/plugins/ds_mongo.c

```diff
@@ -0,0 +1,3126 @@
+/* arrays for elements of the YANG module (without user-ordered elements) */
+/* array for storing nodes with create operation */
+    struct mongo_diff_inner_data cre;
+
+/* array for storing nodes with delete operation */
+    struct mongo_diff_inner_data del;
+
+/* array for storing nodes with replace operation */
+    struct mongo_diff_inner_data rep;
+
+/* array for storing selectors for replace operation */
+    struct mongo_diff_inner_data rep_keys;
+};
+
+static void
+terminate(void)
+{
+    mongoc_cleanup();
+}
+
+static const char *
+ds2str(sr_datastore_t ds)
+{
+    switch (ds) {
+    case SR_DS_STARTUP:
+        return "sr_startup";
+    case SR_DS_RUNNING:
+        return "sr_running";
+    case SR_DS_CANDIDATE:
+        return "sr_candidate";
+    case SR_DS_OPERATIONAL:
+        return "sr_operational";
+    case SR_DS_FACTORY_DEFAULT:
+        return "sr_factory-default";
+    default:
+        return NULL;
+    }
+}
+
+/* try a general command in order to establish whether authentication is required */
+static sr_error_info_t *
+srpds_try_command(mongoc_client_t *client, int *auth_prob)
+{
+    sr_error_info_t *err = NULL;
+    bson_error_t error;
+    bson_t *command = NULL;
+
+    *auth_prob = 0;
+
+    command = BCON_NEW("usersInfo", BCON_INT32(1));
+
+    /* try a command */
+    if (!mongoc_client_command_simple(client, "admin", command, NULL, NULL, &error)) {
+        if (error.code == 11 || error.code == 13) {
+            /* authentication failed | authorization failed */
+            *auth_prob = 1;
+        }
+        else {
+            WRAPERR(&err, SR_ERR_OPERATION_FAILED, "mongoc_client_command_simple()", error.message)
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    bson_destroy(command);
+    return err;
+}
+
+/* establish a database client at the start of the connection */
+static sr_error_info_t *
+srpds_client_init(mongoc_client_pool_t **pool)
+{
+    sr_error_info_t *err = NULL;
+    int auth_prob;
+    mongoc_uri_t *uri = NULL;
+    mongoc_client_t *temp = NULL;
+
+    uri = mongoc_uri_new("mongodb://localhost:27017/?authSource=admin&" MONGOC_URI_SOCKETTIMEOUTMS "=3600000&" MONGOC_URI_CONNECTTIMEOUTMS "=3600000");
+    *pool = mongoc_client_pool_new(uri);
+    temp = mongoc_client_pool_pop(*pool);
+
+    /* try to execute a command without authentication */
+    if ((err = srpds_try_command(temp, &auth_prob))) {
+        goto cleanup;
+    }
+    
+    /* no authentication required, continue */
+    if (!auth_prob) {
+        goto cleanup;
+    }
+
+    /* executing a command without authentication failed */
+    mongoc_client_pool_push(*pool, temp);
+    mongoc_client_pool_destroy(*pool);
+    mongoc_uri_destroy(uri);
+
+    /* authenticate as a client based on the provided username and password */
+    uri = mongoc_uri_new("mongodb://" MONGO_USERNAME ":" MONGO_PASSWORD "@localhost:27017/?authSource=admin&" MONGOC_URI_SOCKETTIMEOUTMS "=3600000&" MONGOC_URI_CONNECTTIMEOUTMS "=3600000");
```

Comment:
Hostname and port should probably be `cmake` variables with these default values.

Reply:
done